### ISSUE TITLE
Would the proof be possible/a lot longer without applies?

### ISSUE THREAD

**Issue by simonzack** [avigad/isabelle#1]
I noticed there are many uses of applies in your proofs of the CLT, e.g. in most of `char_std_normal_distribution`. I'm wondering how feasible it is to write the whole thing in declarative mode, in your experience would it be possible or be a lot longer?

**Comment by johoelzl** [comment]
I'm pretty confident that the proofs can be much smaller and cleaner when written in declarative style. So (1) yes it is possible, and (2) no they might not get longer, quite the contrary, they might get even shorter.

The goal of the current proofs was to finish them and not trying to find a nice structure. I plan to clean them up and especially find nicer more structured proofs in the hope to also find new general lemmas we can add to Isabelle's probability theory. But this depends on how much spare time I have.

**Comment by avigad** [comment]
I am ashamed to admit that many of the proofs that Luke Serafin and I contributed to the proof of the CLT are poorly written. We intended to clean them up after the result was initially achieved, but have since moved on to other projects. Johannes is right that many of the proofs in this repository can be improved, and he has already done a lot of work along those lines.

In general:
- The imperative style is convenient when "hacking" through a proof to get to the conclusion as quickly as possible.
- Imperative proofs are hard to read and hard to maintain.
- Declarative proofs are sometimes longer, because intermediate goals are stated explicitly (this is what makes them easier to read and maintain), but any increase in length for that reason is well worth it.
- The declarative style supports better uses of automation, which can make proofs shorter.

If would want to see examples of nice declarative proofs, I recommend looking at any of Johannes' proofs e.g. in Isabelle's probability theory library, as a model of how things should be done.
